iup/__init__.py

```diff
@@ -0,0 +1,358 @@
+
+    @staticmethod
+    @abc.abstractmethod
+    def split_train_test(uptake_data_list, start_date, side):
+        pass
+
+
+class IncidentUptakeData(UptakeData):
+    def validate(self):
+        self.assert_columns_found(
+            ["region", "season", "date", "elapsed", "interval", "estimate", "previous"]
+        )
+        self.assert_columns_type(["date"], pl.Date)
+        self.assert_columns_type(["region", "season"], pl.Utf8)
+        self.assert_columns_type(
+            ["elapsed", "interval", "estimate", "previous"], pl.Float64
+        )
+
+    @staticmethod
+    def split_train_test(uptake_data_list, start_date, side):
+        if side == "train":
+            out = (
+                pl.concat(uptake_data_list)
+                .sort("date")
+                .filter(pl.col("date") < start_date)
+            )
+        if side == "test":
+            out = (
+                pl.concat(uptake_data_list)
+                .sort("date")
+                .filter(pl.col("date") >= start_date)
+            )
+
+        out = IncidentUptakeData(out)
+        out.validate()
+
+        return out
+
+    def trim_outlier_intervals(self) -> Self:
+        self = self.sort("date").with_row_count("index")
+        unique_regions = self["region"].unique()
+        sub_frames = []
+
+        for x in range(len(unique_regions)):
+            this_region = pl.col("region") == unique_regions[x]
+            if (standardize(self.filter(this_region)["interval"]))[1] > 1:
+                sub_frames.append(self.filter(this_region).slice(3))
+            else:
+                sub_frames.append(self.filter(this_region).slice(2))
+
+        self = pl.concat(sub_frames).sort("date")
+
+        return self
+
+    def to_cumulative(self, last_cumulative):
+        self = (
+            self.select(["region", "date", "estimate"])
+            .sort("date")
+            .with_columns(estimate=pl.col("estimate").cum_sum().over("region"))
+            .join(last_cumulative, on="region")
+            .with_columns(estimate=pl.col("estimate") + pl.col("last_cumulative"))
+            .drop("last_cumulative")
+        )
+
+        self = CumulativeUptakeData(self)
+
+        return self
+
+
+class CumulativeUptakeData(UptakeData):
+    def validate(self):
+        self.assert_columns_found(["region", "date", "estimate"])
+        self.assert_columns_type(["date"], pl.Date)
+        self.assert_columns_type(["region"], pl.Utf8)
+        self.assert_columns_type(["estimate"], pl.Float64)
+        assert (
+            self["estimate"] >= 0.0
+        ).all(), "Cumulative uptake estimates cannot be negative."
+
+    def to_incident(self) -> IncidentUptakeData:
+        self = (
+            self.with_columns(
+                season=(
+                    pl.col("date").dt.year()
+                    + pl.when(pl.col("date").dt.month() < 7).then(-1).otherwise(0)
+                ).cast(pl.Utf8),
+                estimate=pl.col("estimate").diff().over("region"),
```

Comment:
the "over regions" thing is giving me heartburn. Again, if we wanted to add/remove a column (eg population subgroup or demography), then we'd need to go look for the magic word "region" everywhere in the code

I'd much rather having something that's like "grouping variables" or something, and then you are always going "over" those variables, so you only need to have them written in one place

(I'll take this as another opportunity to say that I think it's easier to think about "uptake data" as being *only* dates and estimates. Then all these data frame manipulations become simpler, because you're only looking at a single time series for each one. For example, if you have columns region, season, date, estimate, then:

```
df.partition_by(['region', 'season'], as_dict=True, include_key=False)
```

would give you a dictionary `{(region, season): df}` where each little df has only date & estimate.

I should probably just open a PR to demo this, once this one gets landed.)